python/cudf/cudf/core/algorithms.py

```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+from warnings import warn
+
+import cupy as cp
+
+from cudf.core.series import Index, Series
+
+
+def factorize(values, sort=False, na_sentinel=-1, size_hint=None):
+    """Encode the input values as integer labels
+
+    Parameters
+    ----------
+    values: Series, Index, or CuPy array
+        The data to be factorized.
+    na_sentinel : number, default -1
+        Value to indicate missing category.
+
+    Returns
+    --------
+    (labels, cats) : (Series, Series)
+        - *labels* contains the encoded values
+        - *cats* contains the categories in order that the N-th
+            item corresponds to the (N-1) code.
+
+    Examples
+    --------
+    >>> import cudf
+    >>> data = cudf.Series(['a', 'c', 'c'])
+    >>> codes, uniques = cudf.factorize(data)
+    >>> codes
+    0    0
+    1    1
+    2    1
+    dtype: int8
+    >>> uniques
+    0    a
+    1    c
+    dtype: object
+
+    See Also
+    --------
+    cudf.Series.factorize
+
+    """
+    if sort:
+        raise NotImplementedError(
+            "Sorting not yet supported during factorization."
+        )
+    if na_sentinel is None:
+        raise NotImplementedError("na_sentinel can not be None.")
+
+    if size_hint:
+        warn("size_hint is not applicable for cudf.factorize")
+
+    return_cupy_array = isinstance(values, cp.core.core.ndarray)
+
+    values = Series(values)
+
+    cats = values.dropna().unique().astype(values.dtype)
```

Comment:
Using `Series.dropna` can end up materializing RangeIndexes and generally does unnecessary index handling in general here. Can we avoid it?

Reply:
I think this can be done by just reaching into `._column`. If we have a `Series`, calling `dropna()` will go through `super()` to arrive at the `Frame` level `dropna()` function. From there it ends up falling into `Frame._drop_na_rows()` and eventually to `libcudf.stream_compaction.drop_nulls`, which takes a `libcudf.table.Table` object to operate on. `Table.view` is called next which, since the `Table.index` attribute is a `RangeIndex`, calls `_data` on that range index ultimately materializing the data via `cupy.arange`. 

If we call `column_object.dropnaI()`, then the initial frame is generated using `ColumnBase.as_frame()`. This time, when it reaches `stream_compaction.drop_nulls`, the `_index` attribute is `None`, and so the same is true for the underlying `Table`. As such when it hits `Table.view` there's no materialization.